modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/ValueProvidersParameterDeclarationEnricher.java

```diff
@@ -0,0 +1,266 @@
+import org.mule.runtime.extension.api.declaration.type.ExtensionsTypeLoaderFactory;
+import org.mule.runtime.extension.api.loader.DeclarationEnricher;
+import org.mule.runtime.extension.api.loader.ExtensionLoadingContext;
+import org.mule.runtime.extension.api.values.ValuesProvider;
+import org.mule.runtime.module.extension.internal.loader.java.property.ImplementingTypeModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.ParameterGroupModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.ValuesProviderFactoryModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.ValuesProviderFactoryModelProperty.ValuesProviderFactoryModelPropertyBuilder;
+import org.mule.runtime.module.extension.internal.loader.java.type.ExtensionParameter;
+import org.mule.runtime.module.extension.internal.loader.java.type.FieldElement;
+import org.mule.runtime.module.extension.internal.loader.java.type.runtime.ParameterizableTypeWrapper;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * {@link DeclarationEnricher} implementation that walks through a {@link ExtensionDeclaration} and looks for Source,
+ * Operation, Configuration and Connection Provider Parameters and ParameterGroups annotated with {@link OfValues}.
+ * If a parameter or parameter group is annotated, this one will have a related {@link ValuesProvider}
+ *
+ * @since 4.0
+ */
+public class ValueProvidersParameterDeclarationEnricher extends AbstractAnnotatedDeclarationEnricher {
+
+  private ClassTypeLoader typeLoader;
+
+  @Override
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    Optional<ImplementingTypeModelProperty> implementingType =
+        extractImplementingTypeProperty(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+
+    typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader(currentThread().getContextClassLoader());
+
+    if (implementingType.isPresent()) {
+      new IdempotentDeclarationWalker() {
+
+        @Override
+        public void onSource(SourceDeclaration declaration) {
+          enrichContainerModel(declaration);
+        }
+
+        @Override
+        public void onOperation(OperationDeclaration declaration) {
+          enrichContainerModel(declaration);
+        }
+
+        @Override
+        protected void onConfiguration(ConfigurationDeclaration declaration) {
+          enrichContainerModel(declaration);
+        }
+
+        @Override
+        protected void onConnectionProvider(ConnectionProviderDeclaration declaration) {
+          enrichContainerModel(declaration);
+        }
+      }.walk(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+    }
+
+    typeLoader = null;
```

Comment:
have the typeLoader be a method variable instead of an instance variable, so that you can avoid this and it's less error prone